cache/async_queue.go

```diff
@@ -0,0 +1,61 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+var errAsyncQueueFull = errors.New("the async queue is full")
+
+type asyncQueue struct {
+	stopCh  chan struct{}
+	queueCh chan func()
+
+	workers sync.WaitGroup
+}
+
+func newAsyncQueue(length, maxConcurrency int) *asyncQueue {
+	q := &asyncQueue{
+		stopCh:  make(chan struct{}),
+		queueCh: make(chan func(), length),
+	}
+	// Start a number of goroutines - processing async operations - equal
+	// to the max concurrency we have.
+	q.workers.Add(maxConcurrency)
+	for i := 0; i < maxConcurrency; i++ {
+		go q.asyncQueueProcessLoop()
+	}
+	return q
+}
+
+func (q *asyncQueue) run(op func()) error {
```

Comment:
nit: I would call this `submit()` since it doesn't run the provided closure immediately.

Reply:
Good suggestion!